src/filters/sse.rs

```diff
@@ -52,119 +53,121 @@ use futures::{future, Stream, TryStream, TryStreamExt};
 use http::header::{HeaderValue, CACHE_CONTROL, CONTENT_TYPE};
 use hyper::Body;
 use pin_project::pin_project;
-use serde::Serialize;
-use serde_json;
+use serde_json::{self, Error};
 use tokio::time::{self, Delay};
 
-use self::sealed::{
-    BoxedServerSentEvent, EitherServerSentEvent, SseError, SseField, SseFormat, SseWrapper,
-};
+use self::sealed::SseError;
 use super::header;
 use crate::filter::One;
 use crate::reply::Response;
 use crate::{Filter, Rejection, Reply};
 
-/// Server-sent event message
-pub trait ServerSentEvent: SseFormat + Sized + Send + Sync + 'static {
-    /// Convert to either A
-    fn into_a<B>(self) -> EitherServerSentEvent<Self, B> {
-        EitherServerSentEvent::A(self)
+// Server-sent event data type
+#[derive(Debug)]
+enum DataType {
+    Text(String),
+    Json(String),
+}
+
+/// Server-sent event
+#[derive(Default, Debug)]
+pub struct Event {
+    name: Option<String>,
+    id: Option<String>,
+    data: Option<DataType>,
+    event: Option<String>,
+    comment: Option<String>,
+    retry: Option<Duration>,
+}
+
+impl Event {
+    /// Set Server-sent event data
+    /// data field(s) ("data:<content>")
+    pub fn data<T: Display>(mut self, data: Option<T>) -> Event {
+        self.data = data.map(|d| DataType::Text(d.to_string()));
+        self
     }
 
-    /// Convert to either B
-    fn into_b<A>(self) -> EitherServerSentEvent<A, Self> {
-        EitherServerSentEvent::B(self)
+    /// Set Server-sent event data
+    /// data field(s) ("data:<content>")
+    pub fn json_data<T: Serialize>(mut self, data: Option<T>) -> Result<Event, Error> {
+        self.data = match data {
+            Some(data) => Some(DataType::Json(serde_json::to_string(&data)?)),
+            None => None,
+        };
+        Ok(self)
     }
 
-    /// Convert to boxed
-    fn boxed(self) -> BoxedServerSentEvent {
-        BoxedServerSentEvent(Box::new(self))
+    /// Set Server-sent event comment
+    /// Comment field (":<comment-text>")
+    pub fn comment<T: Display>(mut self, comment: Option<T>) -> Event {
+        self.comment = comment.map(|c| c.to_string());
+        self
     }
-}
 
-impl<T: SseFormat + Send + Sync + 'static> ServerSentEvent for T {}
+    /// Set Server-sent event event
+    /// Event name field ("event:<event-name>")
+    pub fn event<T: Display>(mut self, event: Option<T>) -> Event {
```

Comment:
I might consider changing the bounds to `Into<String>`, since with `T: Display`, passing a `String` will result it creating a 2nd string by calling `to_string()`.

Reply:
changing all the bounds to `Into<String>` breaks cases like the [sse example](https://github.com/seanmonstar/warp/blob/master/examples/sse.rs#L9) should we still do it and update the example, or change just for specific methods?

Reply:
I think it's fine that the example be changed to `data(counter.to_string())` or similar.

Reply:
Question regarding this! Is the string cloned internally, or is the allocation reused for every connection? If it's being cloned, that seems like a _major_ performance problem. Not so much for `fn event`, but for `fn data` or `fn json_data`.